📖 docs: Improve PROJECT file documentation (#3287)
kubernetes-sigs/kubebuilder: docs/book/src/reference/project-config.md

## Overview

> The Project Config represents the configuration of a KubeBuilder project. All projects that are scaffolded with the CLI will generate the `PROJECT` file in the projects' root directory. Therefore, it will store all plugins and input data used to generate the project and APIs to better enable plugins to make useful decisions when scaffolding.

Reviewer: ```suggestion
The Project Config represents the configuration of a KubeBuilder project. All projects that are scaffolded with the CLI (KB version 3.0 and higher) will generate the `PROJECT` file in the projects' root directory. Therefore, it will store all plugins and input data used to generate the project and APIs to better enable plugins to make useful decisions when scaffolding.
```

## Example

> Following an example of an PROJECT file config file which is the result of an project generated with two APIs using the [Deploy Image Plugin][deploy-image-plugin].

Reviewer 2: At the bottom of the page we need to add the alias for this link it is here: https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha.html

So we need the relative path which will be something like 
[deploy-image-plugin]: docs/plugins/deploy-image-plugin-v1-alpha.md 

Please, check the other links in the docs to know how to do that

Author: Added this

> Following an example of an PROJECT file config file which is the result of an project generated with two APIs using the [Deploy Image Plugin][deploy-image-plugin].

Reviewer: ```suggestion
Following is an example of a PROJECT config file which is the result of a project generated with two APIs using the [Deploy Image Plugin][deploy-image-plugin].
```

domain: testproject.org
layout:
- go.kubebuilder.io/v4
plugins:
  deploy-image.go.kubebuilder.io/v1-alpha:
    resources:
    - domain: testproject.org
      group: example.com
      kind: Memcached
      options:
        containerCommand: memcached,-m=64,-o,modern,-v
        containerPort: "11211"
        image: memcached:1.4.36-alpine
        runAsUser: "1001"
      version: v1alpha1
    - domain: testproject.org
      group: example.com
      kind: Busybox
      options:
        image: busybox:1.28
      version: v1alpha1
projectName: project-v4-with-deploy-image
repo: sigs.k8s.io/kubebuilder/testdata/project-v4-with-deploy-image
resources:
- api:
    crdVersion: v1
    namespaced: true
  controller: true
  domain: testproject.org
  group: example.com
  kind: Memcached
  path: sigs.k8s.io/kubebuilder/testdata/project-v4-with-deploy-image/api/v1alpha1
  version: v1alpha1
  webhooks:
    validation: true
    webhookVersion: v1
- api:
    crdVersion: v1
    namespaced: true
  controller: true
  domain: testproject.org
  group: example.com
  kind: Busybox
  path: sigs.k8s.io/kubebuilder/testdata/project-v4-with-deploy-image/api/v1alpha1
  version: v1alpha1
version: "3"

> ## Why do we need to store the plugins and data used?

Reviewer 3: Seems like we need to close the code block above by ` ``` ` ?

Because the sample layout seems incorrect:
https://deploy-preview-3287--kubebuilder.netlify.app/reference/project-config.html?highlight=project#project-config

Author: Sure done that

> Kubebuilder allows users to scaffold the project is external plugins and the PROJECT file can track the information required for the same, for example: 

Reviewer: ```suggestion
Kubebuilder allows users to scaffold the project with external plugins and the PROJECT file can track the information required for the same, for example: 
```

> Kubebuilder allows users to scaffold the project with external plugins and the PROJECT file can track the information required for the same, for example: 

Reviewer 2: ```suggestion
Following some examples of motivations to track the input used: 
```

> - we can check if one plugin can or cannot be scaffolded on top of the existing plugin (ie) plugin compatibility.

Reviewer: ```suggestion
- we can check if a plugin can or cannot be scaffolded on top of an existing plugin (ie) plugin compatibility while chaining multiple of them together.
```

> KubeBuilder allows and provides you opt-in to use plugins, indeed optional ones. Track the information required, for example: 
> - we can check if one plugin can or not be used/supported in the scaffold done
> - what operations can or not be done
> - what data can or not be used in the CLI operations

Reviewer: ```suggestion
Kubebuilder allows users to scaffold the project is external plugins and the PROJECT file can track the information required for the same, for example: 
- we can check if one plugin can or cannot be scaffolded on top of the existing plugin (ie) plugin compatibility.
- what operations can or cannot be done for ex. can multiple APIs be scaffolded for the current configuration or not.
- what data can or not be used in the CLI operations for ex. does an API exists in the project before scaffolding webhooks for the same.
```

> - we can check if a plugin can or cannot be scaffolded on top of an existing plugin (ie) plugin compatibility while chaining multiple of them together.

Reviewer 2: ```suggestion
- check if a plugin can or cannot be scaffolded on top of an existing plugin (i.e.) plugin compatibility while chaining multiple of them together.
```

> - what operations can or cannot be done for ex. can multiple APIs be scaffolded for the current configuration or not.

Reviewer 2: ```suggestion
- what operations can or cannot be done for example; check if multiple APIs can be scaffolded for the current configuration or not.
```

> - what data can or not be used in the CLI operations for ex. does an API exists in the project before scaffolding webhooks for the same.

Reviewer: ```suggestion
- what data can or not be used in the CLI operations for ex. does an API exist in the project before scaffolding webhooks for the same.
```

> - what data can or not be used in the CLI operations for ex. does an API exist in the project before scaffolding webhooks for the same.

Reviewer 2: ```suggestion
- verify what data can or not be used in the CLI operations such as to ensure that WebHooks can only be created for pre-existent API(s)
```

> Note that KubeBuilder is not only a CLI tool but can also be used as a lib to allow others to create their plugins/tools and provide helpers and customizations on top of what is done for it, such as it is done by [Operator-SDK](https://sdk.operatorframework.io/). SDK leverages KubeBuilder to create plugins to allow users to work with other languages and provide helpers for their users to integrate their projects with the [Operator Framework solutions/OLM](https://olm.operatorframework.io/). Check [here](https://book.kubebuilder.io/plugins/creating-plugins.html#why-use-the-kubebuilder-style) to know more about the plugin's vision.

Reviewer 2: Please 
```suggestion
Note that KubeBuilder is not only a CLI tool but can also be used as a lib to allow others to create their plugins/tools and provide helpers and customizations on top of what is done for it, such as it is done by [Operator-SDK](https://sdk.operatorframework.io/). SDK leverages KubeBuilder to create plugins to allow users to work with other languages and provide helpers for their users to integrate their projects with the [Operator Framework solutions/OLM](https://olm.operatorframework.io/). You can check the [plugin's vision](https://book.kubebuilder.io/plugins/creating-plugins.html#why-use-the-kubebuilder-style) to know more about.
```

Please, change all links to alias as you can check the other ones.

> Note that KubeBuilder is not only a CLI tool but can also be used as a lib to allow others to create their plugins/tools and provide helpers and customizations on top of what is done for it, such as it is done by [Operator-SDK](https://sdk.operatorframework.io/). SDK leverages KubeBuilder to create plugins to allow users to work with other languages and provide helpers for their users to integrate their projects with the [Operator Framework solutions/OLM](https://olm.operatorframework.io/). You can check the [plugin's vision](https://book.kubebuilder.io/plugins/creating-plugins.html#why-use-the-kubebuilder-style) to know more about.

Reviewer: ```suggestion
Note that KubeBuilder is not only a CLI tool but can also be used as a library to allow users to create their plugins/tools, provide helpers and customizations on top of their existing projects - an example of which is [Operator-SDK](https://sdk.operatorframework.io/). SDK leverages KubeBuilder to create plugins to allow users to work with other languages and provide helpers for their users to integrate their projects with the [Operator Framework solutions/OLM](https://olm.operatorframework.io/). You can check the [plugin's documentation](https://book.kubebuilder.io/plugins/creating-plugins.html#why-use-the-kubebuilder-style) to know more about creating custom plugins.
```

> Note that KubeBuilder is not only a CLI tool but can also be used as a library to allow users to create their plugins/tools, provide helpers and customizations on top of their existing projects - an example of which is [Operator-SDK](https://sdk.operatorframework.io/). SDK leverages KubeBuilder to create plugins to allow users to work with other languages and provide helpers for their users to integrate their projects with the [Operator Framework solutions/OLM](https://olm.operatorframework.io/). You can check the [plugin's documentation](https://book.kubebuilder.io/plugins/creating-plugins.html#why-use-the-kubebuilder-style) to know more about creating custom plugins.

Reviewer 2: ```suggestion
Note that KubeBuilder is not only a CLI tool but can also be used as a library to allow users to create their plugins/tools, provide helpers and customizations on top of their existing projects - an example of which is [Operator-SDK](https://sdk.operatorframework.io/). SDK leverages KubeBuilder to create plugins to allow users to work with other languages and provide helpers for their users to integrate their projects with, for example, the [Operator Framework solutions/OLM][olm]. You can check the [plugin's documentation][plugins-doc] to know more about creating custom plugins.
```

> On top of that, another motivation for the PROJECT file would help us to create a feature to allow users easier upgrade their projects by providing allowing to automatically re-scaffold the project: Since we scaffold all data used to do the scaffold, we could automate the steps and have a command that would re-create the project in the latest version by re-doing all commands. [More info](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/helper_to_upgrade_projects_by_rescaffolding.md)

Reviewer: ```suggestion
Additionally, another motivation for the PROJECT file is to help us to create a feature that allows users to easily upgrade their projects by providing helpers that automatically re-scaffold the project. By having all the required metadata regarding the APIs, their configurations and versions in the PROJECT file, we could automate the process of re-scaffolding while migrating between plugin versions. More info on this, can be found [here] (https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/helper_to_upgrade_projects_by_rescaffolding.md)
```

> Additionally, another motivation for the PROJECT file is to help us to create a feature that allows users to easily upgrade their projects by providing helpers that automatically re-scaffold the project. By having all the required metadata regarding the APIs, their configurations and versions in the PROJECT file, we could automate the process of re-scaffolding while migrating between plugin versions. More info on this, can be found [here] (https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/helper_to_upgrade_projects_by_rescaffolding.md)

Reviewer 2: ```suggestion
Additionally, another motivation for the PROJECT file is to help us to create a feature that allows users to easily upgrade their projects by providing helpers that automatically re-scaffold the project. By having all the required metadata regarding the APIs, their configurations and versions in the PROJECT file. For example, it can be used to automate the process of re-scaffolding while migrating between plugin versions. ([More info][doc-design-helper]).
```

> The Project Config represents the configuration of a Kubebuilder project. All projects that are scaffolded with the CLI will generate the `PROJECT` file in the projects' root directory.

Reviewer 2: ```suggestion
The Project Config represents the configuration of a KubeBuilder project. All projects that are scaffolded with the CLI will generate the `PROJECT` file in the projects' root directory. Therefore, it will store all plugins and input data used to generate the project and APIs. 
```

> The PROJECT file was introduced since Kubebuilder [release 3.0.0](https://github.com/kubernetes-sigs/kubebuilder/releases/tag/v3.0.0), where the Plugins design was implemented. When scaffolding an operator, the CLI will generate a PROJECT file in the project's root directory. It tracks all data used to perform scaffolds and stores information on resources and plugins in use. These data enable plugins to make more informed decisions during scaffolding, improving their usefulness ([see an example PROJECT file](https://github.com/kubernetes-sigs/kubebuilder/blob/6f1f8c43cfbf260c971037ff039cca86a0980006/testdata/project-v3-with-deploy-image/PROJECT#L2-L21)). 

Reviewer 2: I do not think when it was introduced is relevant
Also, I think is better we use the term projects instead of operators in the docs. 
(Not necessary all projects built with the tool will follow the Operator design pattern)

> The PROJECT file was introduced since Kubebuilder [release 3.0.0](https://github.com/kubernetes-sigs/kubebuilder/releases/tag/v3.0.0), where the Plugins design was implemented. When scaffolding an operator, the CLI will generate a PROJECT file in the project's root directory. It tracks all data used to perform scaffolds and stores information on resources and plugins in use. These data enable plugins to make more informed decisions during scaffolding, improving their usefulness ([see an example PROJECT file](https://github.com/kubernetes-sigs/kubebuilder/blob/6f1f8c43cfbf260c971037ff039cca86a0980006/testdata/project-v3-with-deploy-image/PROJECT#L2-L21)). 

Reviewer 2: ```suggestion
<aside class="note">
<h1>Why do we need to store the plugins and data used?</h1>

KubeBuilder allows and provides you opt-in to use plugins, indeed optional ones. Track the information required, for example: 
- we can check if one plugin can or not be used/supported in the scaffold done
- what operations can or not be done
- what data can or not be used in the CLI operations

Note that KubeBuilder is not only a CLI tool but can also be used as a lib to allow others to create their plugins/tools and provide helpers and customizations on top of what is done for it, such as it is done by [Operator-SDK](https://sdk.operatorframework.io/). SDK leverages KubeBuilder to create plugins to allow users to work with other languages and provide helpers for their users to integrate their projects with the [Operator Framework solutions/OLM](https://olm.operatorframework.io/). Check [here](https://book.kubebuilder.io/plugins/creating-plugins.html#why-use-the-kubebuilder-style) to know more about the plugin's vision.

On top of that, another motivation for the PROJECT file would help us to create a feature to allow users easier upgrade their projects by providing allowing to automatically re-scaffold the project: Since we scaffold all data used to do the scaffold, we could automate the steps and have a command that would re-create the project in the latest version by re-doing all commands. [More info](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/helper_to_upgrade_projects_by_rescaffolding.md)

</>

```

> ## Versioning

Reviewer 2: Could we ensure that we have here the info shared in the following email of this thread? https://groups.google.com/g/kubebuilder/c/5c4_Xc3Fjx0/m/TERhvqaPBQAJ

Also, could we use alias for the docs so that is easier for we keep it maintained?

The links should be at the bottom of the doc

I added some suggestions above ^

Author: Sure I've done the changes as you've recommended

Reviewer 2: Hi @ashutosh887 it still not done, see the following suggestions: 
- (to make links using alias) https://github.com/kubernetes-sigs/kubebuilder/pull/3287/files#r1141269463
- (to add alias at the bottom) https://github.com/kubernetes-sigs/kubebuilder/pull/3287/files#r1141269536

Could you please ensure that all links are with alias and the alias are at the bottom?

## Example

> Following an example of an PROJECT file config file which is the result of an project generated with two APIs using the [Deploy Image Plugin](../plugins/deploy-image-plugin-v1-alpha.md).

Reviewer 2: We need to use alias.
We let all links/paths at the bottom of the pages see all docs. 

That make easier we keep the things maintained

Author: Got it. Updated it in the next commit @camilamacedo86

[project]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3/PROJECT
[versioning]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/VERSIONING.md#Versioning
[core-types]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/golang/options.go

> [deploy-image-plugin]: ../plugins/deploy-image-plugin-v1-alpha.md

Reviewer 2: ```suggestion
[deploy-image-plugin]: ../plugins/deploy-image-plugin-v1-alpha.md
[olm]: https://olm.operatorframework.io/
[plugins-doc]: ../plugins/creating-plugins.html#why-use-the-kubebuilder-style
```
